<a href="https://colab.research.google.com/github/n1az/Intro-to-deep-learning-with-keras/blob/main/Week%205/Final%20Assignment/Peer-graded%20Assignment%3A%20Build%20a%20Regression%20Model%20in%20Keras%20(D).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Download and Clean Dataset

Let's start by importing the <em>pandas</em> and the Numpy libraries.

In [22]:
import pandas as pd
import numpy as np

We will be using the dataset provided in the assignment

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

<strong>1. Cement</strong>

<strong>2. Blast Furnace Slag</strong>

<strong>3. Fly Ash</strong>

<strong>4. Water</strong>

<strong>5. Superplasticizer</strong>

<strong>6. Coarse Aggregate</strong>

<strong>7. Fine Aggregate</strong>

Let's read the dataset into a <em>pandas</em> dataframe.

In [23]:
concrete_data = pd.read_csv('sample_data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Fine Aggregate  Age  Strength
0   540.0                 0.0      0.0  ...           676.0   28     79.99
1   540.0                 0.0      0.0  ...           676.0   28     61.89
2   332.5               142.5      0.0  ...           594.0  270     40.27
3   332.5               142.5      0.0  ...           594.0  365     41.05
4   198.6               132.4      0.0  ...           825.5  360     44.30

[5 rows x 9 columns]

So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

#### Let's check how many data points we have.

In [24]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

Let's check the dataset for any missing values.

In [25]:
concrete_data.describe()

Cement  Blast Furnace Slag  ...          Age     Strength
count  1030.000000         1030.000000  ...  1030.000000  1030.000000
mean    281.167864           73.895825  ...    45.662136    35.817961
std     104.506364           86.279342  ...    63.169912    16.705742
min     102.000000            0.000000  ...     1.000000     2.330000
25%     192.375000            0.000000  ...     7.000000    23.710000
50%     272.900000           22.000000  ...    28.000000    34.445000
75%     350.000000          142.950000  ...    56.000000    46.135000
max     540.000000          359.400000  ...   365.000000    82.600000

[8 rows x 9 columns]

In [26]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

The data looks very clean and is ready to be used to build our model.

#### Split data into predictors and target

The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

In [27]:
concrete_data_columns = concrete_data.columns
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

Let's do a quick sanity check of the predictors and the target dataframes.

In [28]:
predictors.head()

Cement  Blast Furnace Slag  Fly Ash  ...  Coarse Aggregate  Fine Aggregate  Age
0   540.0                 0.0      0.0  ...            1040.0           676.0   28
1   540.0                 0.0      0.0  ...            1055.0           676.0   28
2   332.5               142.5      0.0  ...             932.0           594.0  270
3   332.5               142.5      0.0  ...             932.0           594.0  365
4   198.6               132.4      0.0  ...             978.4           825.5  360

[5 rows x 8 columns]

In [29]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation.

In [30]:
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

Cement  Blast Furnace Slag  ...  Fine Aggregate       Age
0  2.476712           -0.856472  ...       -1.217079 -0.279597
1  2.476712           -0.856472  ...       -1.217079 -0.279597
2  0.491187            0.795140  ...       -2.239829  3.551340
3  0.491187            0.795140  ...       -2.239829  5.055221
4 -0.790075            0.678079  ...        0.647569  4.976069

[5 rows x 8 columns]

In [31]:
n_cols = predictors_norm.shape[1] # number of predictors
n_cols

8

<a id="item1"></a>

<a id="item1"></a>

## Import Keras

#### Let's go ahead and import the Keras library

In [32]:
import keras

As you can see, the TensorFlow backend was used to install the Keras library.

Let's import the rest of the packages from the Keras library that we will need to build our regressoin model.

In [33]:
from keras.models import Sequential
from keras.layers import Dense

In [50]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_cols,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function creates a model that has one hidden layer with 10 neurons and a ReLU activation function. It uses the adam optimizer and the mean squared error as the loss function.

Let's import scikit-learn in order to randomly split the data into a training and test sets

In [51]:
from sklearn.model_selection import train_test_split

Splitting the data into a training and test sets by holding 30% of the data for testing

In [52]:
X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=42)

## Train and Test the Network

Let's call the function now to create our model.

In [53]:
# build the model
model = regression_model()

Next, we will train the model for 50 epochs.


In [54]:
# fit the model
epochs = 50
model.fit(X_train, y_train, epochs=epochs, verbose=2)

Epoch 1/50
23/23 - 1s - loss: 1616.2944
Epoch 2/50
23/23 - 0s - loss: 1591.5151
Epoch 3/50
23/23 - 0s - loss: 1565.0837
Epoch 4/50
23/23 - 0s - loss: 1527.0908
Epoch 5/50
23/23 - 0s - loss: 1469.7101
Epoch 6/50
23/23 - 0s - loss: 1381.3126
Epoch 7/50
23/23 - 0s - loss: 1247.7563
Epoch 8/50
23/23 - 0s - loss: 1060.2009
Epoch 9/50
23/23 - 0s - loss: 836.5757
Epoch 10/50
23/23 - 0s - loss: 608.5220
Epoch 11/50
23/23 - 0s - loss: 426.8763
Epoch 12/50
23/23 - 0s - loss: 324.1215
Epoch 13/50
23/23 - 0s - loss: 275.0564
Epoch 14/50
23/23 - 0s - loss: 246.5112
Epoch 15/50
23/23 - 0s - loss: 227.4789
Epoch 16/50
23/23 - 0s - loss: 214.6201
Epoch 17/50
23/23 - 0s - loss: 205.4584
Epoch 18/50
23/23 - 0s - loss: 198.1575
Epoch 19/50
23/23 - 0s - loss: 192.4610
Epoch 20/50
23/23 - 0s - loss: 187.4932
Epoch 21/50
23/23 - 0s - loss: 183.4084
Epoch 22/50
23/23 - 0s - loss: 180.3029
Epoch 23/50
23/23 - 0s - loss: 176.5038
Epoch 24/50
23/23 - 0s - loss: 173.4391
Epoch 25/50
23/23 - 0s - loss: 170.8800
E

Next we need to evaluate the model on the test data.

In [55]:
loss_val = model.evaluate(X_test, y_test)
y_pred = model.predict(X_test)
loss_val

10/10 [==============================] - 0s 1ms/step - loss: 134.4886


134.48858642578125

Now we need to compute the mean squared error between the predicted concrete strength and the actual concrete strength.

Let's import the mean_squared_error function from Scikit-learn.

In [56]:
from sklearn.metrics import mean_squared_error

In [57]:
mean_square_error = mean_squared_error(y_test, y_pred)
mean = np.mean(mean_square_error)
standard_deviation = np.std(mean_square_error)
print(mean, standard_deviation)

134.48858393384847 0.0


Create a list of 50 mean squared errors and report mean and the standard deviation of the mean squared errors.

In [59]:
total_mean_squared_errors = 50
epochs = 50
mean_squared_errors = []
for i in range(0, total_mean_squared_errors):
    X_train, X_test, y_train, y_test = train_test_split(predictors_norm, target, test_size=0.3, random_state=i)
    model.fit(X_train, y_train, epochs=epochs, verbose=0)
    MSE = model.evaluate(X_test, y_test, verbose=0)
    print("MSE "+str(i+1)+": "+str(MSE))
    y_pred = model.predict(X_test)
    mean_square_error = mean_squared_error(y_test, y_pred)
    mean_squared_errors.append(mean_square_error)

mean_squared_errors = np.array(mean_squared_errors)
mean = np.mean(mean_squared_errors)
standard_deviation = np.std(mean_squared_errors)

print('\n')
print("Below is the mean and standard deviation of " +str(total_mean_squared_errors) + " mean squared errors with normalized data. Total number of epochs for each training is: " +str(epochs) + "\n")
print("Mean: "+str(mean))
print("Standard Deviation: "+str(standard_deviation))

MSE 1: 32.67182159423828
MSE 2: 33.5646858215332
MSE 3: 32.52947998046875
MSE 4: 31.892353057861328
MSE 5: 28.350250244140625
MSE 6: 30.225046157836914
MSE 7: 31.70294189453125
MSE 8: 26.885019302368164
MSE 9: 25.741649627685547
MSE 10: 26.51274299621582
MSE 11: 27.047138214111328
MSE 12: 24.907875061035156
MSE 13: 29.888368606567383
MSE 14: 32.1361083984375
MSE 15: 25.90570068359375
MSE 16: 21.740219116210938
MSE 17: 27.33246421813965
MSE 18: 26.16060447692871
MSE 19: 25.981456756591797
MSE 20: 27.486305236816406
MSE 21: 23.172231674194336
MSE 22: 26.189695358276367
MSE 23: 24.10103416442871
MSE 24: 25.95882225036621
MSE 25: 24.850675582885742
MSE 26: 29.258697509765625
MSE 27: 20.576128005981445
MSE 28: 24.834341049194336
MSE 29: 29.876726150512695
MSE 30: 26.46396255493164
MSE 31: 25.36017417907715
MSE 32: 21.043312072753906
MSE 33: 22.62589454650879
MSE 34: 25.151792526245117
MSE 35: 25.5878963470459
MSE 36: 27.981252670288086
MSE 37: 19.885116577148438
MSE 38: 24.749710083007812
M

So we got mean value 25.885 which is less than the previous value 30.963 and std is now 3.456 which was 10.5 previously.